lib/maths/CBoostedTreeLoss.cc

```diff
@@ -205,6 +223,212 @@ CArgMinLogisticImpl::TDoubleVector CArgMinLogisticImpl::value() const {
     result(0) = minimum;
     return result;
 }
+
+CArgMinMultinomialLogisticLossImpl::CArgMinMultinomialLogisticLossImpl(std::size_t numberClasses,
+                                                                       double lambda,
+                                                                       const CPRNG::CXorOShiro128Plus& rng)
+    : CArgMinLossImpl{lambda}, m_NumberClasses{numberClasses}, m_Rng{rng},
+      m_ClassCounts{TDoubleVector::Zero(numberClasses)}, m_PredictionSketch{NUMBER_CENTRES} {
+}
+
+std::unique_ptr<CArgMinLossImpl> CArgMinMultinomialLogisticLossImpl::clone() const {
+    return std::make_unique<CArgMinMultinomialLogisticLossImpl>(*this);
+}
+
+bool CArgMinMultinomialLogisticLossImpl::nextPass() {
+
+    using TMeanAccumulator = CBasicStatistics::SSampleMean<TDoubleVector>::TAccumulator;
+
+    if (m_CurrentPass == 0) {
+        TKMeans::TSphericalClusterVecVec clusters;
+        if (m_PredictionSketch.kmeans(NUMBER_CENTRES, clusters) == false) {
+            m_Centres.push_back(TDoubleVector::Zero(m_NumberClasses));
+            m_CurrentPass += 2;
+        } else {
+            // Extract the k-centres.
+            m_Centres.reserve(clusters.size());
+            for (const auto& cluster : clusters) {
+                TMeanAccumulator centre{TDoubleVector::Zero(m_NumberClasses)};
+                for (const auto& point : cluster) {
+                    centre.add(point);
+                }
+                m_Centres.push_back(CBasicStatistics::mean(centre));
+            }
+            std::sort(m_Centres.begin(), m_Centres.end());
+            m_Centres.erase(std::unique(m_Centres.begin(), m_Centres.end()),
+                            m_Centres.end());
+            LOG_TRACE(<< "# centres = " << m_Centres.size());
+            m_CurrentPass += m_Centres.size() == 1 ? 2 : 1;
+            m_CentresClassCounts.resize(m_Centres.size(),
+                                        TDoubleVector::Zero(m_NumberClasses));
+        }
+
+        // Reclaim the memory used by k-means.
+        m_PredictionSketch = TKMeans{0};
+    } else {
+        ++m_CurrentPass;
+    }
+
+    LOG_TRACE(<< "current pass = " << m_CurrentPass);
+
+    return m_CurrentPass < 2;
+}
+
+void CArgMinMultinomialLogisticLossImpl::add(const TMemoryMappedFloatVector& prediction,
+                                             double actual,
+                                             double weight) {
+
+    using TMinAccumulator = CBasicStatistics::SMin<std::pair<double, std::size_t>>::TAccumulator;
+
+    switch (m_CurrentPass) {
+    case 0: {
+        // We have a member variable to avoid allocating a tempory each time.
+        m_DoublePrediction = prediction;
+        m_PredictionSketch.add(m_DoublePrediction, weight);
```

Comment:
The basic idea is as follows. We want to find the `weight = argmin_w { sum_{i in I}{ [-log([softmax(prediction_i + w)]_{actual_i}]) } + lambda * w'w }`. Here, `I` is the set of all training examples in the leaf and `actual_i` is the index of the i'th example actual category. Rather than working with this function directly we summarise it by the set `{(x_j, c_j)}` where `x_j` are some points in prediction space and `c_j` are the counts of the nearest predictions in the `I` to each `x_j`. We use a k-means of the predictions in `I` to choose `x_j` this is calculated sequentially (to accommodate the case we're using disk storage). This is what `m_PredictionSketch` is doing. I'll add some class documentation to explain the strategy.